### PR TITLE
C#: Extract default parameter values.

### PR DESCRIPTION
**Pull request by michaelnebel**
It turns out that there is yet another way, one can define a *parameter default value* by using the attribute (`DefaultParameterValue`) to instruct the compiler, how the default value should be constructed using an implicit conversion operator.

The following is valid C# code:
```csharp
public struct MyStruct
{
    public static implicit operator MyStruct(string s) => new MyStruct();
}

public class MyClass
{
    public void M([Optional, DefaultParameterValue("mystring")] MyStruct arg) => throw null;
}
```
The compiler checks that there exists a way to construct (in this case an implicit conversion operator) a value of type `MyStruct` from a `string`.
Intuitively, the definition of `M` is equivalent to (which is not legal C# code).
```csharp
public void M(MyStruct arg = "mystring") => throw null;
```

In this PR we try to emulate this behaviour.

### CODE REVIEW

**Review by hvitved — APPROVED** [review]
Looks great to me 🎉

**Comment by hvitved** [comment]
Did you run DCA already?

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/ImplicitCast.cs [hunk 1]
```diff
@@ -11,33 +12,74 @@ public Expression Expr
...
+        private static IMethodSymbol? GetImplicitConversionMethod(ITypeSymbol type, object value) =>
+            type
+                .GetMembers()
+                .Where(m =>
+                    m is IMethodSymbol method &&
+                    method.GetName() == "op_Implicit" &&
+                    method.Parameters.Length == 1 &&
+                    method.Parameters[0].Type.Name == value.GetType().Name
+                )
+                .Cast<IMethodSymbol>()
```
Comment (tamasvajk) [inline csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/ImplicitCast.cs:51]: You might be able to simplify this `.GetMembers().Where(m => m is IMethodSymbol ...).Cast<IMethodSymbol>()` to `.GetMembers().OfType<IMethodSymbol>().Where(...)`

**Review by tamasvajk — APPROVED** [review]

File: csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/ImplicitCast.cs (hunk 1 above)
Comment (michaelnebel) [reply csharp/extractor/Semmle.Extraction.CSharp/Entities/Expressions/ImplicitCast.cs:51]: Yes, thank you! I will do that :-)

**Comment by michaelnebel** [comment]
> Did you run DCA already?

Not yet - but I will do so :-)

**Comment by michaelnebel** [comment]
DCA looks good!

**Review by tamasvajk — APPROVED** [review]

**Pull request merged**